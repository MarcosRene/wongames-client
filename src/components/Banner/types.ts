export type BannerProps = {
  img: string
  title: string
  subtitle: string
  buttonLabel: string
  buttonLink: string
  ribbon?: React.ReactNode
  ribbonColor?: 'primary' | 'secondary'
  ribbonSize?: 'small' | 'normal'
}
