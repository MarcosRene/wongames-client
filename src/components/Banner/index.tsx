import Image from 'next/image'

import Button from 'components/Button'
import Ribbon from 'components/Ribbon'

import { BannerProps } from './types'

import * as S from './styles'

const Banner = ({
  img,
  title,
  subtitle,
  buttonLabel,
  buttonLink,
  ribbon,
  ribbonColor = 'primary',
  ribbonSize = 'normal'
}: BannerProps) => (
  <S.Wrapper>
    {!!ribbon && (
      <Ribbon color={ribbonColor} size={ribbonSize}>
        {ribbon}
      </Ribbon>
    )}

    <S.ImageWrapper>
      <Image src={img} alt={title} layout="fill" objectFit="cover" />
    </S.ImageWrapper>

    <S.Caption>
      <S.Title>{title}</S.Title>
      <S.Subtitle dangerouslySetInnerHTML={{ __html: subtitle }} />
      <Button as="a" href={buttonLink} size="large">
        {buttonLabel}
      </Button>
    </S.Caption>
  </S.Wrapper>
)

export default Banner
