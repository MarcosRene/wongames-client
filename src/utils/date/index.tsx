export default function formatedDate(releaseDate: string) {
  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(new Date(releaseDate))
}
