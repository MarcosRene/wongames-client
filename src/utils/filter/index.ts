import { ParsedUrlQueryInput } from 'querystring'
import { ItemProps } from 'components/ExploreSidebar/types'

type ParseArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: Pick<ItemProps, 'type' | 'name'>[]
}

export function parseQueryStringToFilters({
  queryString,
  filterItems
}: ParseArgs) {
  const obj: Record<string, string> = {}

  for (const key of Object.keys(queryString).filter(
    (item) => item !== 'sort'
  )) {
    const item = filterItems.find((filterItem) => filterItem.name === key)
    const isCheckbox = item?.type === 'checkbox'

    obj[key] = (
      !isCheckbox
        ? queryString[key]
        : {
            name: queryString[key]
          }
    ) as string
  }

  return obj
}

export function parseQueryStringToFilter({
  queryString,
  filterItems
}: ParseArgs) {
  const obj: Record<string, string[]> = {}

  for (const key of Object.keys(queryString)) {
    const type = filterItems.find((filterItem) => filterItem.name === key)?.type
    const isCheckbox = type === 'checkbox'
    const isArray = Array.isArray(queryString[key])

    obj[key] =
      !isArray && isCheckbox
        ? [queryString[key] as string]
        : (queryString[key] as string[])
  }

  return obj
}
