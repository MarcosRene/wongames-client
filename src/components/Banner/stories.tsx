import { Meta, StoryObj } from '@storybook/react'
import styled from 'styled-components'

import Banner from '.'

export default {
  title: 'Banner',
  args: {
    img: 'https://source.unsplash.com/user/willianjusten/1042x580',
    title: 'Defy death',
    subtitle: '<p>Play the new <strong>CrashLands</strong> season',
    buttonLabel: 'Buy now',
    buttonLink: '/games/defy-death'
  },
  argTypes: {
    ribbon: {
      type: 'string'
    }
  },
  component: Banner
} as Meta<typeof Banner>

const Container = styled.div`
  max-width: 104rem;
  margin: 0 auto;
`

export const Default: StoryObj<typeof Banner> = {
  render: (args) => (
    <Container>
      <Banner {...args} />
    </Container>
  )
}

export const WithRibbon: StoryObj<typeof Banner> = {
  render: (args) => (
    <Container>
      <Banner {...args} />
    </Container>
  )
}

WithRibbon.args = {
  ribbon: '20% OFF',
  ribbonSize: 'normal',
  ribbonColor: 'primary'
}
