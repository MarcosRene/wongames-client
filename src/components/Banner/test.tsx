import { screen } from '@testing-library/react'

import { renderWithTheme } from 'utils/tests'
import theme from 'styles/theme'

import Banner from '.'

const bannerProps = {
  img: 'https://source.unsplash.com/user/willianjusten/1042x580',
  title: 'Defy death',
  subtitle: '<p>Play the new <strong>CrashLands</strong> season',
  buttonLabel: 'Buy now',
  buttonLink: '/games/defy-death'
}

describe('<Banner />', () => {
  it('should render correctly', () => {
    const { container } = renderWithTheme(<Banner {...bannerProps} />)

    expect(
      screen.getByRole('heading', { name: /defy death/i })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('heading', { name: /play the new crashLands season/i })
    ).toBeInTheDocument()

    expect(screen.getByRole('img', { name: /defy death/i })).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a Ribbon', () => {
    renderWithTheme(
      <Banner
        {...bannerProps}
        ribbon="My Ribbon"
        ribbonSize="small"
        ribbonColor="secondary"
      />
    )

    const ribbon = screen.getByText(/My Ribbon/i)

    expect(ribbon).toBeInTheDocument()
    expect(ribbon).toHaveStyle({ backgroundColor: theme.colors.secondary })
    expect(ribbon).toHaveStyle({
      height: '2.6rem',
      fontSize: '1.2rem'
    })
  })
})
