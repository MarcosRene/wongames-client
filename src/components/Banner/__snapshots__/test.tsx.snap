// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
.c5 {
  padding: 0.8rem;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

.c5:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c0 {
  position: relative;
}

.c1 {
  width: 100%;
  height: 23rem;
  background-color: #EAEAEA;
  position: relative;
}

.c2 {
  width: 100%;
  padding: 2.4rem;
  background-color: rgba(0,0,0,0.7);
}

.c3 {
  font-size: 1.8rem;
  font-weight: 600;
  color: #FAFAFA;
}

.c4 {
  margin-bottom: 1.6rem;
  color: #FAFAFA;
  font-size: 1.4rem;
  font-weight: 400;
}

.c4 strong {
  font-weight: 600;
  color: #F231A5;
}

@media (max-width:1170px) {

}

@media (min-width:768px) {
  .c0 {
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0,0,0,0.2);
  }
}

@media (min-width:768px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c2 {
    position: absolute;
    bottom: 0;
    left: 0;
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 4.0rem;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 2.8rem;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 1.8rem;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <span
      style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
    >
      <img
        alt="Defy death"
        data-nimg="fill"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
      />
      <noscript />
    </span>
  </div>
  <div
    class="c2"
  >
    <h2
      class="c3"
    >
      Defy death
    </h2>
    <h3
      class="c4"
    >
      <p>
        Play the new 
        <strong>
          CrashLands
        </strong>
         season
      </p>
    </h3>
    <a
      class="c5"
      href="/games/defy-death"
    >
      <span>
        Buy now
      </span>
    </a>
  </div>
</main>
`;
